pytrickle/frames.py

```diff
@@ -179,4 +244,235 @@ def with_monotonic_timestamps(cls, frames: List[AudioFrame], request_id: str, st
+        return FrameProcessor.create_processed_frame(processed_tensor, original_frame)
+    
+    def convert_trickle_audio_to_av(self, frame: AudioFrame) -> av.AudioFrame:
+        """Convert pytrickle AudioFrame to av.AudioFrame."""
+        return FrameProcessor.convert_trickle_audio_to_av(frame)
+    
+    def convert_av_audio_to_trickle(self, av_frame: av.AudioFrame, original_frame: AudioFrame) -> AudioFrame:
+        """Convert av.AudioFrame back to pytrickle AudioFrame with original timing."""
+        return FrameProcessor.convert_av_audio_to_trickle(av_frame, original_frame)
+
+
+# Streaming Utilities
+# ====================
+
+class FrameBuffer:
+    """Rolling frame buffer that keeps a fixed number of frames."""
+    
+    def __init__(self, max_frames: int = 300):
+        self.max_frames = max_frames
+        self.frames: Deque[Union[VideoFrame, AudioFrame]] = deque(maxlen=max_frames)
+        self.total_frames_received = 0
+        self.total_frames_discarded = 0
+        
+    def add_frame(self, frame: Union[VideoFrame, AudioFrame]):
+        if len(self.frames) >= self.max_frames:
+            self.total_frames_discarded += 1
+        self.frames.append(frame)
+        self.total_frames_received += 1
+        
+    def get_frame(self) -> Optional[Union[VideoFrame, AudioFrame]]:
+        return self.frames.popleft() if self.frames else None
+        
+    def get_all_frames(self) -> List[Union[VideoFrame, AudioFrame]]:
+        frames = list(self.frames)
+        self.frames.clear()
+        return frames
+        
+    def clear(self):
+        self.frames.clear()
+        
+    def size(self) -> int:
+        return len(self.frames)
+        
+    def get_stats(self) -> Dict[str, int]:
+        return {
+            "current_frames": len(self.frames),
+            "max_frames": self.max_frames,
+            "total_received": self.total_frames_received,
+            "total_discarded": self.total_frames_discarded
+        }
+
+
+class StreamState:
+    """Unified state management for stream lifecycle."""
+    
+    def __init__(self):
+        self.running = False
+        self.pipeline_ready = False
+        self.shutting_down = False
+        self.error_occurred = False
+        self.cleanup_in_progress = False
+        
+        self.running_event = asyncio.Event()
+        self.shutdown_event = asyncio.Event()
+        self.error_event = asyncio.Event()
+        self.pipeline_ready_event = asyncio.Event()
+    
+    @property
+    def is_active(self) -> bool:
+        return self.running and not self.shutting_down and not self.error_occurred
+    
+    @property
+    def shutdown_flags(self) -> Dict[str, bool]:
+        return {
+            'shutdown_event': self.shutdown_event.is_set(),
+            'cleanup_in_progress': self.cleanup_in_progress
+        }
+    
+    def start(self):
+        self.running = True
+        self.running_event.set()
+    
+    def mark_pipeline_ready(self):
```

Comment:
I think this makes more sense but not overly opinionated about it.
```suggestion
    def pipeline_ready(self):
```